### ISSUE TITLE
Enhancement : Respect STRIP_STYLE project option

### ISSUE THREAD

**Issue by dive** [Carthage/Carthage#2485]
* carthage install method: [ ] .pkg, [X] homebrew, [ ] source
* `which carthage`:  `/usr/local/bin/carthage`
* `carthage version`: `0.29.0`
* `xcodebuild -version`:  `Xcode 9.4 Build version 9F1027a`
* Are you using `--no-build`?: No
* Are you using `--no-use-binaries`?: No
* Are you using `--use-submodules`?: No
* Are you using `--cache-builds`?: No
* Are you using `--new-resolver`?: No

**Cartfile**
```
No cart file. We use the following command to build:
carthage build --platform ios --no-skip-current $verbose_arg
```

**Carthage Output**
```
-
```

**Actual outcome**
Carthage did not respect `STRIP_STYLE = non-global` option for the project during the build because of [the following code](https://github.com/Carthage/Carthage/blob/1295134ce2694aa60107398d2f83d136b3e86398/Source/CarthageKit/Xcode.swift#L716):
```
Carthage/Source/CarthageKit/Xcode.swift:716
...
// Disable the "Strip Linked Product" build
// setting so we can later generate a dSYM
"STRIP_INSTALLED_PRODUCT=NO",
...
```
As a consequence, the `strip` within Carthage is executed without `-x` option and in some cases, we get a huge unexpected binary. Of course, we can `strip` it later with a proper `strip` options, but it's an overhead (`.dSYM` becomes useless, needs time to execute, etc.). Also, these `frameworks` cannot be stripped **properly** within application target (for Release, etc.) because we do not have the information about `STRIP_STYLE` setting for each **prebuilt** `framework` at this point.

**Expected outcome**
Carthage should(?) respect `STRIP_STYLE` option for `frameworks`, call `strip` with proper options and skip `.dSYM` generation if necessary. From my point of view, `.xcconfig` itself (or project settings) must be the source of truth.
At least, let's show a warning message or information note during the build if Carthage is going to ignore some inherited options.

**Comment by blender** [comment]
You should be using `carthage copy-frameworks`, that will strip you binaries during copy

**Comment by dive** [comment]
Hey @blender, it shouldn't and doesn't. Because of two things:

- the information about `STRIP_STYLE` [is unavailable](https://github.com/Carthage/Carthage/blob/master/Source/carthage/CopyFrameworks.swift#L118) for **prebuilt** `frameworks` on this stage because it's usually called within application build phases as I mentioned in the ticket;
```
private func shouldStripDebugSymbols() -> Bool {
	return getEnvironmentVariable("COPY_PHASE_STRIP")
		.map { $0 == "YES" }.value ?? false
}
```
- also, check [.stripDebugSymbols](https://github.com/Carthage/Carthage/blob/1295134ce2694aa60107398d2f83d136b3e86398/Source/CarthageKit/Xcode.swift#L1073) implementation. It uses only predefined values for `strip`:
```
arguments: [ "strip", "-S", "-o", binaryURL.path, binaryURL.path]
```
But for `STRIP_STYLE = non-global` we have to use `-x` option as mentioned in the man page for `strip`:
>  -x     Remove all local symbols (saving only global symbols).

**Comment by jdhealy** [comment]
@dive, you’re probably aware, but just for [context on the decision to set `STRIP_INSTALLED_PRODUCT` to `NO`](https://github.com/Carthage/Carthage/pull/2334#issue-164416797), quoting @abrindam:

> This change is necessary [since Carthage moved to using the "archive" command](https://github.com/Carthage/Carthage/pull/2158) to build. Most products have `STRIP_INSTALLED_PRODUCT` set to true, but this does not kick in unless you have `DEPLOYMENT_POSTPROCESSING` set to true as well, and that is false for most products. However, using "archive" instead of "build" magically overrides `DEPLOYMENT_POSTPROCESSING` to true, resulting in the product being stripped of debug symbols. As a result dSYM generation produces effectively no information for builds that are built with "archive" (e.g. OSX and iOS device, but not iOS simulator)
> 
> This also restores the 0.25 behavior that built frameworks are not stripped. Prior to [fix](https://github.com/Carthage/Carthage/pull/2334), stripping was happening with the archive command, which may have caused additional issues that were not reported. With this change, we are back to the old behavior and stripping is done at framework copy time.

[The decision of what flags to include in `stripDebugSymbols` did consider the `-x` flag](https://github.com/Carthage/Carthage/pull/2361#discussion_r171566088), but a resolution wasn't (at that point) found that could bridge the gap between `STRIP_STYLE` being known at compile time and having to later strip downloads of that compilation product.

**Comment by blender** [comment]
> But for STRIP_STYLE = non-global we have to use -x

as to why `-x` is missing:

https://github.com/Carthage/Carthage/pull/2361#issuecomment-370429346
https://github.com/Carthage/Carthage/pull/2361#issuecomment-370436508

@dive 

Do you have a suggestion on how to retain dSYM generation and solve your problem? \

Maybe add parameters to copy-frameworks?

**Comment by dive** [comment]
Hey @jdhealy. Thanks for the excellent resume on the issue. Yeah, I'm aware of these conversations that's why I've raised this question again as an enhancement. 

I like the idea of `setxattr`. The solution looks good but a little unobvious that probably does not matter in this case. 

Of course, we can try a brute force solution and embed a simple generated `.xconfig` into the `framework` on Carthage build phase with all necessary information about overridden options and then read this file on `copy-frameworks` to configure `strip` options properly. We'll have to remove this `.xconfig` from the `framework` as well before `code_sign` but it's not an issue at to me. But I like the idea of `setxattr`much more.

@blender, the problem is that we cannot add parameters to `copy-frameworks` blindly. We have to restore `strip` options for each `framework` before applying. See the proposed solution with `setxattr` above.

**Comment by jdhealy** [comment]
> But we need to be sure that `setxattr` are stored in Git properly.

Not quite straight from the source, but [extended attributes are not encoded into `git` repositories](http://git.661346.n2.nabble.com/Revisiting-metadata-storage-td6991047.html).

But, consider: users — if made aware that `xattr`s are being written — can choose to workaround: by using zips from `binary` or GitHub releases, by documenting it in team guidelines, by forking Carthage, even by not building with Carthage and just using it a resolver and checkout tool.

So, we could make users aware with output when `xattr`s will be written, and include in the output a hyperlink to a page outlining workaround options.

If that seems too noisy, we could gate by certain calendar date after which it no longer gets output.

Both the line about the change and also that hyperlink to a page outlining workarounds would, of course, be included in [the release notes](https://github.com/Carthage/Carthage/releases).

**Comment by dive** [comment]
> But, consider: users — if made aware that xattrs are being written — can choose to workaround: by using zips from `binary` or GitHub releases, by documenting it in team guidelines, by forking Carthage, even by not building with Carthage and just using it a resolver and checkout tool.

Yeah, but at least it will not solve the issue with local pre-built binaries. For example, we build one of our frameworks with Carthage locally and store it within the repository. Right now we have a custom build script that applies our custom `strip` attributes, but it would be nice to have a universal solution.

> So, we could make users aware with output when xattrs will be written, and include in the output a hyperlink to a page outlining workaround options.

It would be more informative if we will notify users about overridden build options by Carthage during the build phase as to me.

I think we can implement an additional parameter for Carthage that will preserve all user-defined build options inside a framework and will use this information during `copy-frameworks` if its available. The parameter will be optional and will not affect the current behaviour for Carthage. Something like `carthage build --preserve-build-options`. What do you think, @jdhealy?

Also a bit of philosophy :) As @blender mentioned above while discussing the brute force solution with simple embedded `.xcconfig` to restore proper attributes in `copy-frameworks`:

> I'm not 100% this is acceptable given carthage's philosophy...

But, from my point of view, the current behaviour contradicts the following postulate described in ["Differences between Carthage and CocoaPods"](https://github.com/Carthage/Carthage#differences-between-carthage-and-cocoapods) because we override user-defined options explicitly and without any notice:

> Ultimately, we created Carthage because we wanted the simplest tool possible—a dependency manager that gets the job done **without taking over the responsibility of Xcode**, and **without creating extra work for framework authors**.

**Comment by blender** [comment]
Ok, let me take one step back here

> It would be more informative if we will notify users about overridden build options by Carthage during the build phase as to me.

As far and I understand the only override Carthage is doing is `STRIP_INSTALLED_PRODUCT=NO` to get dSYMs generated. This in turn disables whatever strip options you have set because they are not applied.

>This change is necessary since Carthage moved to using the "archive" command to build. Most products have STRIP_INSTALLED_PRODUCT set to true, but this does not kick in unless you have DEPLOYMENT_POSTPROCESSING set to true as well, and that is false for most products. However, using "archive" instead of "build" magically overrides DEPLOYMENT_POSTPROCESSING to true, resulting in the product being stripped of debug symbols. As a result dSYM generation produces effectively no information for builds that are built with "archive" (e.g. OSX and iOS device, but not iOS simulator)

So, are we sure there is no way to get the dSYM generated here?

@dive Can you try to archive the framework in question and check if it's stripped by Xcode as you expect and if the dSYM is generated? I have the suspicion that one of the two will not happen.


I am open to anything but imho:
- adding yet another CLI switch is confusing as there are quite a few already
- `xattr` won't survive long
- emdebbing an extra `.xcconfig` in people's frameworks is too me too big of an intrusion


Technically we can also just generate another file that sits along side the frameworks in `/Carthage/Build/...` where we can shove all sorts of meta. I also though about (ab)using the .version file but it seems wrong to me.

**Comment by blender** [comment]
ping

**Comment by dive** [comment]
@blender, hey! Sorry for the delay but I have no time to check it right now. Maybe next week. At this stage, we decided to switch back to `xcodebuild` for our needs because there is too much "magic" here. But I will return to this problem as soon as I have free time and we will think of something. Also, I'm going to check your theories about `STRIP_INSTALLED_PRODUCT`, `DEPLOYMENT_POSTPROCESSING` and `dSYM` generations.

P.S. also I've noticed that Carthage ignores `DEBUG_INFORMATION_FORMAT = dwarf` option and generates `dSYM` even when it's explicitly noted that we do not need them for the module.

**Comment by abrindam** [comment]
@sheehanm You're absolutely right, in my opinion. Using "archive" does not prevent dSYM generation. 

I'm the one who implemented the overriding of setting to facilitate dSYM generation in it's current form. When I did the implementation, I explicitly pointed out there were two directions we could go:

1. Suppress the stripping defaults of "archive" and generate our own dSYM
2. Use the dSYM generated by the "archive" command.

The response was that people wanted the library built by Carthage to always have symbols, regardless of the library's own build settings. So we went with #1.

See https://github.com/Carthage/Carthage/issues/2280#issuecomment-356825509

I think you make a persuasive argument that we should have gone in the other direction, to be honest. Specifically, Carthage has always valued simplicity and avoided magic and we might be trending too far away from that here.

I do think we really need to answer the fundamental question, which is, "do users expect the frameworks build by Carthage to be stripped of symbols, even if the author of that framework has explicitly enabled symbol stripping?" I think we can really only move forward if we answer that question.

**Comment by blender** [comment]
> Why does Carthage takes it upon itself to override things that should be controlled by the build settings?

You can read yourself the motivation for the overrides.

https://github.com/Carthage/Carthage/blob/49c7901f851384a358bac2e5d2333cabc4f91900/Source/CarthageKit/Xcode.swift#L695-L716

> Carthage has always valued simplicity and avoided magic and we might be trending too far away from that here.

I agree.

So a potential fix to this would be to run https://github.com/Carthage/Carthage/blob/master/Source/CarthageKit/Xcode.swift#L588 only if `DEPLOYMENT_POSTPROCESSING` is `NO` ?

BTW Almofire has `DEPLOYMENT_POSTPROCESSING=NO` and the value does not change while archiving 

https://github.com/Carthage/Carthage/issues/2280#issuecomment-356825509
>  However, DEPLOYMENT_POSTPROCESSING seems to be magically true when you use archive, leading to stripping.

**Comment by blender** [comment]
I just tried and I think we're on the right path here. I'll try to make a branch so that you guys can check.

**Comment by blender** [comment]
here is the branch: https://github.com/blender/Carthage/tree/fix/2485-respect-strip-style

Don't rely on this in production because. It is not the final version and it would need some heavy reworking to make everything work as expected.

**Comment by saijithendrag** [comment]
Is this something that can help with reducing framework sizes ? We are having nearly 50 dependencies in our project and start of this month iPA size was 77 MB and end of this month it became 159 MB. We inspected iPA files and all this size is because of frameworks increased by 3 times in span of 6-7 test builds. And this increase is only for Carthage built frameworks. Any idea what could happen ?

**Comment by blender** [comment]
> Is this something that can help with reducing framework sizes ?

Yes

> We inspected iPA files and all this size is because of frameworks increased by 3 times in span of 6-7 test builds.

Not all symbols are stripped. If you have specified a different strip style other than -S (debug symbols) then you could be experiencing a binary size increase. 

Changes have happened over 3 versions of Carthage (0.28.0, 0.29.0 and 0.30.1). I don't know how often you build but I'm fairly certain binary size should be constant given a specific version of Carthage .

**Comment by saijithendrag** [comment]
We build almost everyday and it keeps on increasing the binary size with new builds since start of this month. Our app size doubled in 1 month just with this framework sizes and no idea what was wrong or what we had done different to cause this. We are on 0.30.1 version of Carthage.

<img width="524" alt="screen shot 2018-07-30 at 2 36 29 pm" src="https://user-images.githubusercontent.com/4617560/43416429-3548e37e-9406-11e8-9b60-42e549d53097.png">

**Comment by blender** [comment]
I think we're all on the same page here acknowledging that the more code is in your frameworks the bigger the size? 

That said, if the strip style of your frameworks is "All Symbols" carthage unfortunately does not respect that at the moment. It only strips debug symbols during `carthage copy-frameworks` if Strip Debug Symbols During Copy a.k.a COPY_PHASE_STRIP=YES is set in the project where you use `carthage copy-frameworks`.

So then, in your app set "Strip Debug Symbols During Copy" to yes.

If that is not enough, you have to strip your frameworks manually for all symbols at the moment.

**Comment by acecilia** [comment]
@tmspzz Thanks for your work, I see your branch [here](https://github.com/blender/Carthage/tree/fix/2485-respect-strip-style) has the basics to fix this.

You mentioned that:

> It is not the final version and it would need some heavy reworking to make everything work as expected

Specifically for this issue:
* > if DEPLOYMENT_POSTPROCESSING=NO no dSYMS are generated at all

I believe is not a problem. Consumers can use `export XCODE_XCCONFIG_FILE=...`before calling carthage to set `DEPLOYMENT_POSTPROCESSING = YES` if they need.

What else do you think is needed?

**Comment by joshuarendora1** [comment]
Before we process let me ask you a little favor to please Subscribe this
Youtube Channel. Link 👇👇👇

https://youtube.com/shorts/cEVQqoR-Qas?feature=share

Thank You So Much.

Let me know if your done ♥️

**Comment by tmspzz** [comment]
@acecilia I am afraid I don't recall, it's been 5 years now. 

The changes should be reevaluated in the light of .xcframeworks as well.

**Comment by joshuarendora1** [comment]
Hi Good Day before we begin can i ask a little favor to Subscribe these
Youtube chanel?

Please click Link Below. 👇👇


https://youtube.com/shorts/va3bZK4wiKU?feature=share



YOU'RE SUBSCRIPTION IS REALLY A BIG HELP TO THE CHARITY WE CHOSEN TO
RECEIVED 75% OF THE SALLARY WE MAKE IN THE FUTURE.

WE ARE EXPECTING THAT YOU ARE ALSO IN LINE AND 1 OF THOSE PEOPLE ALSO WHO
SUPPORT WHAT WE MAKE AND PLAN IN THE FUTURE IN HELPING OTHERS ♥️♥️♥️

THANK YOU SO MUCH ♥️
